core/src/test/scala/neotypes/BaseIntegrationSpec.scala

```diff
@@ -1,43 +1,57 @@
 package neotypes
 
 import java.time.Duration
-
 import com.dimafeng.testcontainers.{ForAllTestContainer, GenericContainer}
 import org.neo4j.driver.{v1 => neo4j}
+import org.scalatest.flatspec.AsyncFlatSpecLike
 import org.testcontainers.containers.wait.strategy.HostPortWaitStrategy
-import org.scalatest.flatspec.AsyncFlatSpec
+import scala.concurrent.Future
 
-/** Base class for simple integration specs. */
-abstract class BaseIntegrationSpec[F[_]] extends AsyncFlatSpec with ForAllTestContainer {
-  def initQuery: String
+/** Base class for writing integration specs. */
+abstract class BaseIntegrationSpec[F[_]](testkit: EffectTestkit[F]) extends BaseEffectSpec(testkit) with AsyncFlatSpecLike with ForAllTestContainer {
+  protected def initQuery: String
 
-  override val container = GenericContainer("neo4j:3.5.3",
+  override final val container = GenericContainer("neo4j:3.5.3",
     env = Map("NEO4J_AUTH" -> "none"),
     exposedPorts = Seq(7687),
     waitStrategy = new HostPortWaitStrategy().withStartupTimeout(Duration.ofSeconds(30))
   )
 
-  lazy val driver = neo4j.GraphDatabase.driver(s"bolt://localhost:${container.mappedPort(7687)}")
+  protected lazy final val driver =
+    neo4j.GraphDatabase.driver(s"bolt://localhost:${container.mappedPort(7687)}")
 
-  def execute[T](work: Session[F] => F[T])
-                (implicit F: Async[F]): F[T] =
-    (new Driver[F](driver)).writeSession(work)
+  protected lazy final val session =
+    driver.session()
 
-  override def afterStart(): Unit = {
-    if (initQuery != null) {
-      val session = driver.session()
-      try {
-        session.writeTransaction(
-          new neo4j.TransactionWork[Unit] {
-            override def execute(tx: neo4j.Transaction): Unit =
-              tx.run(initQuery)
-          }
-        )
-      } finally {
-        session.close()
+  private final def runQuery(query: String): Unit = {
+    session.writeTransaction(
+      new neo4j.TransactionWork[Unit] {
+        override def execute(tx: neo4j.Transaction): Unit =
+          tx.run(query)
       }
+    )
+  }
+
+  override final def afterStart(): Unit = {
+    if (initQuery != null) {
+      runQuery(initQuery)
```

Comment:
CleaningIntegrationSpec doesn't need an initQuery value, but it must inherit one from BaseIntegrationSpec.  BaseIntegrationSpec defines a cleanDb method, but it is only used in CleaningIntegrationSpec.  Could these be broken out into a different OO hierarchy, maybe CleanIntegrationSpec and DirtyIntegrationSpec?

Reply:
> CleaningIntegrationSpec doesn't need an initQuery value

Well, it does, an empty one.

> BaseIntegrationSpec defines a cleanDb method, but it is only used in CleaningIntegrationSpec

Yeah, I believe I can move that method into **CleaningIntegrationSpec**.

> Could these be broken out into a different OO hierarchy

That would just make things more complicated.
A **CleaningIntegrationSpec** is an **IntegrationSpec** that start all tests with an empty db.

Reply:
Ah I remember, I create `cleanDb` on the **BaseIntegrationSpec**. Because otherwise, I would have to expose the `runQuery` method, which is very powerful.
Also, someday `cleanDb` may be useful for other specs.

So I would leave this as it is.

Reply:
👍